rfcs/032-calm-deletions/README.md

```diff
@@ -0,0 +1,30 @@
+# RFC 032: Calm deletion watcher
+
+**Status:** Draft
+
+**Last updated:** 28/01/2021
+
+## Background
+
+We've realised that the pipeline does not pick up deleted Calm records - once they're deleted there is no trace of them in the Calm API, but they remain in the source VHS tables. Especially because a cleanup project at the end of last year was responsible for a very large number of deletions, this needs to be resolved.
+
+So far as we can tell, there is no way to find deleted records other than checking for their absence. This means that our solution needs to take some sort of polling-based approach: going through the source table and checking for the continued presence of each record in the Calm API.
+
+## Proposal
+
+- The deletion watcher will be an ECS task within the existing calm adapter cluster.
+- It will run as a scheduled task at a frequency to be agreed with the collections information team.
+- It will also be possible to run the task manually via the `RunTask` API. A script will be provided for this.
+- The task will scan the source data VHS table and check for the existence of each record in the Calm API. If the record has been deleted (is no longer present) then it will update the VHS entry to flag the deletion. 
+- The calm transformer will check for the presence of the deletion flag in the source data and create `Deleted` works as appropriate.
+
+This architecture is illustrated below (arrows indicate the direction of data flow):
+
+![image](https://user-images.githubusercontent.com/4429247/106134857-08103780-615f-11eb-845f-112109ead907.png)
+
+## Questions & potential issues
+
+- Are we really 100% sure that polling for deletions is the only way to detect them?
+- Can we reuse any existing things for this (eg the reindexer)?
```

Comment:
We can reuse some of the reindexer code, in particular the parallel scan. That allows you to distribute work among many workers, rather than having a single worker that does the whole table.

Or just have the reindexer fire records directly at whatever's polling Calm.

Reply:
I think this task should stay within calm adapter infra rather than using the reindexer. I'm not sure if you mean multiple workers in the sense of multiple tasks but I'm not sure this needs to scale to > 1 task, if it's only running once a week or something like that it doesn't matter if it takes a little while...

Reply:
> I think this task should stay within calm adapter infra rather than using the reindexer

Why? We already cross the cluster boundary elsewhere.

> I'm not sure if you mean multiple workers in the sense of multiple tasks but I'm not sure this needs to scale to > 1 task, if it's only running once a week or something like that it doesn't matter if it takes a little while...

True, and we don't want to hammer the Calm API too hard, but see comment above about using SQS to handle being interrupted halfway through.

Reply:
Given that the job of the adapters is to reflect the true source data inside our walls, I think the deletion watcher is very clearly a part of the calm adapter - so if we were to do this I think we'd want to put the reindex worker task inside the adapter cluster (which is perfectly achievable) so it would all be deployed into one place.

Reply:
I’m not sure I agree – wouldn’t that mean we should put the reindex in the catalogue pipeline clusters as well? Or indeed, smush everything into one super cluster?

Reply:
I don't think so - you have to draw the boundaries somewhere and that's where I believe they should be for this work. As I say, it doesn't actually make much difference here as we can run the reindexer task wherever we want.

Reply:
So I guess I'm just wary of having multiple instances of the reindex worker floating around, because it adds complexity for deployment and triggering a reindex.

It's been acceptable to have the reindex worker living separately from the tasks before, and inter-cluster communication over queues. I'm still struggling to see the benefit, and what makes this process different from the rest of the pipeline.

Reply:
Hmm, I guess it's because this wouldn't be doing a reindex - but perhaps the reindexer is already far more overloaded than I thought it was.